agent/consul/rpc.go

```diff
@@ -443,6 +445,157 @@ RUN_QUERY:
 	return err
 }
 
+type sharedQueryFn func(memdb.WatchSet, *state.Store) (uint64, func(uint64, interface{}) error, error)
+
+func (s *Server) sharedBlockingQuery(req cache.Request, res interface{}, queryOpts *structs.QueryOptions, queryMeta *structs.QueryMeta, fn sharedQueryFn) error {
+	var timeout *time.Timer
+
+	if queryOpts.RequireConsistent {
+		if err := s.consistentRead(); err != nil {
+			return err
+		}
+	}
+
+	// Restrict the max query time, and ensure there is always one.
+	if queryOpts.MaxQueryTime > maxQueryTime {
+		queryOpts.MaxQueryTime = maxQueryTime
+	} else if queryOpts.MaxQueryTime <= 0 {
+		queryOpts.MaxQueryTime = defaultQueryTime
+	}
+
+	// Apply a small amount of jitter to the request.
+	queryOpts.MaxQueryTime += lib.RandomStagger(queryOpts.MaxQueryTime / jitterFraction)
+
+	// Setup a query timeout.
+	timeout = time.NewTimer(queryOpts.MaxQueryTime)
+	defer timeout.Stop()
+
+	cacheInfo := req.CacheInfo()
```

Comment:
I share your feeling, I used this because it did the right thing for this use case and was ready to use. 
This is a naming / namespacing issue to me, both the cache and this PR need an unique identifier for the request parameters. It could be moved out of `cache.Request` interface to have its own method with both the cache and this PR depending on it.